### PR TITLE
torchprep [experimental]

### PR DESCRIPTION
**Pull request by msaroufim**
v0.1 as of now supports
* Quantization based CLI
* Profiling (w/ warmup)
* Using enums as types for api discovery
* Fusion
* Easy build, installation and upload to pypi with poetry
* Setting CPU environment variables on machine where inference is happening
* Sample models BERT and Resnet
* Torchscript fusion
* Pruning

Next steps
* Automatic distillation
* Upload to pypi
* Better ways of setting and recording environment variables for inferencing engines
* Multi tensor input
* See roadmap at bottom of `README.md`

To install `pip install poetry`

After 0.1 we can focus on a good story for environment variables and docker containers required for fast inference with IPEX, TensorRT and others

```
cd torchprep/torchprep
poetry install
python example.py
```

Run command `torchprep quantize models/resnet152.pt int8 --profile 64,3,7,7`

Sample output

```
Model succesfully quantized
model:  base model : Size (MB): 241.669177
model:  quantized_model : Size (MB): 241.669813
Starting profile
Average latency for base model is: 85.4 ms
Min latency for base model is: 83 ms
Max p99 latency for base model is: 98 ms
Starting profile
Average latency for quantized_model is: 47.26 ms
Min latency for quantized_model is: 45 ms
Max p99 latency for quantized_model is: 52 ms
model quantized_model.pt was saved
```

Usage instructions

```
(torchprep) ubuntu@ip-172-31-62-217:~/serve$ torchprep quantize --help
Usage: torchprep quantize [OPTIONS] MODEL_PATH PRECISION:{int8|float16}

  Quantize a saved torch model to a lower precision float format to reduce its
  size and latency

Arguments:
  MODEL_PATH                [required]
  PRECISION:{int8|float16}  [required]

Options:
  --device [cpu|gpu]  [default: cpu]
  --profile TEXT      Comma seperated input tensor shape
  --help              Show this message and exit.

```

```
Usage: torchprep [OPTIONS] COMMAND [ARGS]...

Options:
  --install-completion  Install completion for the current shell.
  --show-completion     Show completion for the current shell, to copy it or
                        customize the installation.
  --help                Show this message and exit.

Commands:
  distill        Create a smaller student model by setting a distillation...
  env-variables
  fuse           Supports optimizations including conv/bn fusion, dropout...
  hello
  quantize       Quantize a saved torch model to a lower precision float...
```

### CODE REVIEW

File: torchprep/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+```
+
+## Usage
+* Float16 precision on GPU `python main.py quantize --precision float16 --device gpu`
+* Float16 quantization on CPU `python main.py quantize --float16`
+* Int8 quantization on CPU profiled with a random matrix of size 2x2: `python main.py quantize --precision int8 --profile [2,2]`
+
+## Coming soon
+* Reduce parameter count by 1/3 `python main.py distil model.pt -distil 1/3`
+* And much more
```
Comment (HamidShojanazeri) [inline torchprep/README.md:18]: Thanks @msaroufim , this is great. As per  the off-line discussion,  TensorRT, ORT support for optimizations also can be added, with considering to add them for third party requirements instead of docker containers.

File: torchprep/torchprep/torchprep/main.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    """
+    model = load_model(model_path, device)
+
+    if device == Device.cpu:
+        if precision == "int8":
+            dtype = torch.qint8
+        else:
+            dtype = torch.float16
+    quantized_model = torch.quantization.quantize_dynamic(
+    model, {torch.nn.LSTM, torch.nn.Linear, torch.nn.Conv2d}, dtype=dtype
```
Comment (jamesr66a) [inline torchprep/torchprep/torchprep/main.py:48]: Cool! I can definitely see this as being profitable for `LSTM` and `Linear` layers. I'm wondering about `Conv2d`. Since the balance of parameters/activations is usually flipped for Conv compared to the other two (conv has larger activations and smaller parameters), dynamic quantization might incur high overhead in CNN cases. Do you have any insight on this?

Comment (jamesr66a) [reply torchprep/torchprep/torchprep/main.py:48]: I see the `profile` stuff below. Is the idea that this is a human-in-the-loop tool and that the user would be checking that the perf is actually better?

**Comment by yqhu** [comment]
@msaroufim It's super cool to do quantization and profiling in one command!  A couple of quick comments:
* peotry is cool but not as widely used as pip; perhaps start with pip to make it easier to install?
* would it be possible to infer the input tensor shapes for common models?
* add warmup iterations in profiling?

**Comment by msaroufim** [comment]
Thank you for the feedback @yqhu!

> poetry is cool but not as widely used as pip; perhaps start with pip to make it easier to install?

Once I upload the library to pypi you'll be able to pip install it without downloading poetry or being aware of it, I'd rather keep poetry for now for the build workflow since it's been orders of magnitude more productive for me

> would it be possible to infer the input tensor shapes for common models?

In principle it should be because if you give a model the wrong shape it'll say expected shape X, but googling online seems like this is not obvious. I'll keep digging and if you have any suggestions please let me know!

> add warmup iterations in profiling?

Makes tons of sense, will add

File: experimental/torchprep/torchprep/main.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    typer.echo("coming soon")
+
+@app.command()
+def env_variables(model_path : Path, architecture : Architecture) -> None:
+    """
+    Set environment variables for optimized inference. Run this command on the machine where inference will happen!
+    """
+    if architecture == Architecture.ipex:
+        os.environ["OMP_NUM_THREADS"] = 1
+        os.environ["KMP_BLOCKTIME"] = 1
```
Comment (lxning) [inline experimental/torchprep/torchprep/main.py:64]: Can these 2 env var be configurable?

File: experimental/torchprep/torchprep/main.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    torch.save(optimized_model, 'optimized_model.pt') 
+    return optimized_model
+    typer.echo("coming soon")
+
+@app.command()
+def env_variables(model_path : Path, architecture : Architecture) -> None:
+    """
+    Set environment variables for optimized inference. Run this command on the machine where inference will happen!
+    """
+    if architecture == Architecture.ipex:
```
Comment (lxning) [inline experimental/torchprep/torchprep/main.py:62]: How to integrate [ipex parameters](https://github.com/intel/intel-extension-for-pytorch/blob/master/tutorials/intro_launch.md#usage-of-launch-script)?

File: experimental/torchprep/torchprep/main.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        duration = toc - tic
+        duration = math.trunc(duration * 1000)
+        durations.append(duration)
+    avg = sum(durations) / len(durations)
+    min_latency = min(durations)
+    max_latency = max(durations)
+    print(f"Average latency for {label} is: {avg} ms")
+    print(f"Min latency for {label} is: {min_latency} ms")
+    print(f"Max p99 latency for {label} is: {max_latency} ms")
+    return [avg, min_latency, max_latency]
```
Comment (lxning) [inline experimental/torchprep/torchprep/main.py:129]: Is it possible to add all parameter settings including env settings into the profile report?

File: experimental/torchprep/torchprep/main.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        if isinstance(module, torch.nn.Conv2d) or isinstance(module, torch.nn.Linear) or isinstance(module, torch.nn.LSTM):
+            torch.nn.utils.prune.l1_unstructured(module, name, prune_amount)
+    
+    torch.save(model, 'pruned_model.pt')
+    return model
+
+
+
+@app.command()
+def fuse(model_path : Path, device : Device = Device.cpu,input_shape : str = typer.Option(default=None, help="Comma seperated input tensor shape")) -> torch.nn.Module:
```
Comment (HamidShojanazeri) [inline experimental/torchprep/torchprep/main.py]: @msaroufim we might be able to use a more descriptive name here.

File: experimental/torchprep/torchprep/main.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+            torch.nn.utils.prune.l1_unstructured(module, "weight", prune_amount)
+    
+    torch.save(model, 'pruned_model.pt')
+    print("Saved prune model pruned_model.pt")
+    return model
+
+
+
+@app.command()
+def fuse(model_path : Path, device : Device = Device.cpu,input_shape : str = typer.Option(default=None, help="Comma seperated input tensor shape e.g [64,3,7,7]"), input_type : TensorType = TensorType.float) -> torch.nn.Module:
```
Comment (HamidShojanazeri) [inline experimental/torchprep/torchprep/main.py:61]: @msaroufim we might be able to use a more descriptive name here, similar to optimize_for_inference?

File: experimental/torchprep/torchprep/main.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    min_latency = min(durations)
+    max_latency = max(durations)
+    print(f"Average latency for {label} is: {avg} ms")
+    print(f"Min latency for {label} is: {min_latency} ms")
+    print(f"Max p99 latency for {label} is: {max_latency} ms")
+    return [avg, min_latency, max_latency]
+
+def load_model(model_path: str, device="cpu") -> torch.nn.Module:
+    map_location = torch.device(device)
+    model = torch.load(model_path, map_location=map_location)
```
Comment (HamidShojanazeri) [inline experimental/torchprep/torchprep/main.py:185]: @msaroufim should we cover loading  [state_dict/ torchscript](https://github.com/pytorch/serve/blob/master/ts/torch_handler/base_handler.py#L96) as well? or even make use existing functions?

File: experimental/torchprep/README.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+```sh
+torchprep quantize --help
+```
+
+### Example
+
+```sh
+# Download resnet example
+python example.py
```
Comment (HamidShojanazeri) [inline experimental/torchprep/README.md:48]: @msaroufim seems transformers are missing from requirements, it has import error, and torchvision

File: experimental/torchprep/README.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+# quantize a cpu model with int8 on cpu and profile with a float tensor of shape [64,3,7,7]
+torchprep quantize models/resnet152.pt int8 --input-shape 64,3,7,7
```
Comment (HamidShojanazeri) [inline experimental/torchprep/README.md:51]: @msaroufim as we discussed, it would be helpful to clarify on tensor type here.

File: experimental/torchprep/README.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+# profile a model for a 100 iterations
+torchprep profile models/resnet152.pt --iterations 100 --device cpu --input-shape 64,3,7,7
+
+# set omp threads to 1 to optimize cpu inference
+torchprep env --device cpu
+
+# Prune 30% of model weights
+torchprep prune models/resnet152.pt --prune_amount 0.3
```
Comment (HamidShojanazeri) [inline experimental/torchprep/README.md:60]: @msaroufim typos,  prune-amount

Comment (HamidShojanazeri) [reply experimental/torchprep/README.md:60]: it would be good to add the profile for prune as well

File: experimental/torchprep/torchprep/main.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        toc = time.time()
+        duration = toc - tic
+        duration = math.trunc(duration * 1000)
+        durations.append(duration)
+    avg = sum(durations) / len(durations)
+    min_latency = min(durations)
+    max_latency = max(durations)
+    print(f"Average latency for {label} is: {avg} ms")
+    print(f"Min latency for {label} is: {min_latency} ms")
+    print(f"Max p99 latency for {label} is: {max_latency} ms")
```
Comment (HamidShojanazeri) [inline experimental/torchprep/torchprep/main.py:180]: how about p99 "np.quantile(durations, 0.99)"

**Review by HamidShojanazeri — APPROVED** [review]
Thanks @msaroufim, I think this is a great start and this can evolve to cover many more optimizations. 

I left some minor comments.

**Review by lxning — APPROVED** [review]
